packages/core/src/lib/tooltip/tooltip-styles.ts

```diff
@@ -27,48 +49,92 @@ export interface TooltipStyles {
   };
 }
 
-export const tooltipStyles = (): Styles => {
-  const { subscribe, set } = writable<TooltipStyles>({
-    tooltip: {},
-    arrow: {},
-  });
-
-  const recalculate = async (
-    targetElement: HTMLElement | undefined,
-    tooltipElement: HTMLElement | undefined,
-    arrowElement: HTMLElement | undefined,
-    location: TooltipLocation
-  ) => {
-    if (targetElement && tooltipElement && arrowElement) {
-      const nextStyles = await calculateStyle(
-        targetElement,
-        tooltipElement,
-        arrowElement,
-        location
-      );
-      set(nextStyles);
-    }
+export interface State {
+  location?: TooltipLocation;
+  visibility?: TooltipVisibility;
+  target?: HTMLElement | undefined;
+  tooltip?: HTMLElement | undefined;
+  arrow?: HTMLElement | undefined;
+  isHovered?: boolean;
+}
+
+const CONTEXT_KEY = Symbol('tooltip');
+const INITIAL_STYLE: Readonly<TooltipStyles> = { tooltip: {}, arrow: {} };
+
+export const provideTooltipContext = (): TooltipContext => {
+  const context = createContext();
+
+  setContext(CONTEXT_KEY, context);
+
+  return context;
+};
+
+export const useTooltip = (): TooltipContext => {
+  const context = getContext<TooltipContext | undefined>(CONTEXT_KEY);
+
+  if (!context) {
+    throw new Error('Usage: tooltip context required');
+  }
+
+  return context;
+};
+
+const createContext = (): TooltipContext => {
+  const id = useUniqueId('tooltip');
+  const state = writable<State>({});
+  const isVisible = derived(
+    state,
+    ($state) => $state.visibility === 'visible' || Boolean($state.isHovered)
+  );
+  const styles = derived<Readable<State>, TooltipStyles>(
+    state,
+    ($state, set) => {
+      const { target, tooltip } = $state;
+
+      return target && tooltip
+        ? autoUpdate(target, tooltip, () => updateStyle($state, set))
+        : noop;
+    },
+    INITIAL_STYLE
+  );
```

Comment:
Gonna write this up in the wiki, but this was my first async derived store. It was the right tool for this particular job, but it had enough quirks that I'd be careful about using it in the future